File: Cap 10/10.8 - Recursividade ou Recorrencia/10.8.2 - Recursividade de Cauda/recauda.jl
#=
 * Livro .....: Algoritmos
 *              Logica para Desenvolvimento de Programacao de
 *              Computadores
 * Autores ...: Jose Augusto N. G. Manzano
 *              Jayr Figueiredo de Oliveira
 * Editora ...: Erica
 *
 * Capitulo ..: 10 - Topico 10.8
 * 
 * Programa ..: FUNCAO COM RECURSIVDADE DE CAUDA
=#

function FATORIALBASE(N, P)
  if (N == 0)
    return P
  else
    return FATORIALBASE(N - 1, N * P)
    end
end

function FATORIAL(N)
  return FATORIALBASE(N, 1)
end

println("CALCULO DE FATORIAL")
println("RECURSIVIDADE DE CAUDA")
println()

print("Qual fatorial (1-20): ")
LIMITE = parse(Int, readline())

println("Fatorial = ", FATORIAL(LIMITE))

File: Cap 10/10.8 - Recursividade ou Recorrencia/10.8.2 - Recursividade de Cauda/recauda.vb
' Livro .....: Algoritmos
'              Logica para Desenvolvimento de Programacao de 
'              Computadores
' Autores ...: Jose Augusto N. G. Manzano
'              Jayr Figueiredo de Oliveira
' Editora ...: Erica
' 
' Capitulo ..: 10 - Topico 10.8
'  
' Programa ..: FUNCAO COM RECURSIVDADE DE CAUDA

Module Module1

    Function FATORIALBASE(ByVal N As Byte, ByVal P As ULong) As ULong
        If (N = 0) Then
            FATORIALBASE = P
        Else
            FATORIALBASE = FATORIALBASE(N - 1, N * P)
        End If
    End Function

    Function FATORIAL(ByVal N As Byte) As ULong
        FATORIAL = FATORIALBASE(N, 1)
    End Function

    Sub Main()
  
        Dim LIMITE As Byte

        Console.WriteLine("CALCULO DE FATORIAL")
        Console.WriteLine("RECURSIVIDADE DE CAUDA")
        Console.WriteLine()

        Console.Write("Qual fatorial (1-20): ")
        LIMITE = Byte.Parse(Console.ReadLine())

        Console.WriteLine("Fatorial = " & FATORIAL(LIMITE))
  
    End Sub

End Module
